### PR TITLE
Remove the Cython layer and use ctypes

### LINKED ISSUE #164: make the Python wrapper accept pre-allocated output array
It'd be nice if the Python wrapper would allow the user to pass a pre allocated array for output instead of [allocating ](https://github.com/vais-ral/CCPi-Regularisation-Toolkit/blob/master/src/Python/src/cpu_regularisers.pyx#L115)the output by default.

### LINKED ISSUE #177: NameError raised in tests is not caught, hence test crash if no GPU code is compiled
These lines raise a `NameError` rather than `ValueError`. 
https://github.com/vais-ral/CCPi-Regularisation-Toolkit/blob/fc67dc0db27807d4964af7b83f8ce4460abd9324/test/test_run_test.py#L73

All tests crash, see for instance https://github.com/SyneRBI/SIRF-SuperBuild/actions/runs/5073437582/jobs/9112456133?pr=813#step:10:986

### LINKED ISSUE #178: build error
I'm getting this error when trying to import a module: 
```
   from ccpi.filters.cpu_regularisers import TV_ROF_CPU, TV_FGP_CPU, TV_PD_CPU, TV_SB_CPU, dTV_FGP_CPU, TNV_CPU, NDF_CPU, Diff4th_CPU, TGV_CPU, LLT_ROF_CPU, PATCHSEL_CPU, NLTV_CPU
ImportError: /home/algol/miniconda3/conda-bld/ccpi-regulariser_1694256493616/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold/lib/python3.10/site-packages/ccpi/filters/cpu_regularisers.cpython-310-x86_64-linux-gnu.so: undefined symbol: _Z14SB_TV_CPU_mainPfS_S_fifiiii
```
Seems that it is related to `SB_TV_CPU_main` function, but I cannot find any issues with it. Any ideas @paskino? Many thanks

### PR DESCRIPTION
**Pull request by paskino**
Removes the Cython layer in favour of `ctypes`. My aim has been to generally maintain the user experience however, this change is **backward incompatible**.

- following https://github.com/vais-ral/CCPi-Regularisation-Toolkit/pull/182#issuecomment-1988738648
- closes #178
- closes #164 
- closes #177 
- [x] removes a lot of boilerplate code, but more could be removed. 
- [x] working unit tests
- [x] working [demo CPU vs GPU regularisers](https://github.com/vais-ral/CCPi-Regularisation-Toolkit/blob/use_ctypes/demos/demo_cpu_vs_gpu_regularisers.py)
- [x] working https://github.com/vais-ral/CCPi-Regularisation-Toolkit/blob/use_ctypes/demos/demo_cpu_regularisers3D.py
- [x] working https://github.com/vais-ral/CCPi-Regularisation-Toolkit/blob/use_ctypes/demos/demo_gpu_regularisers.py
- [x] working https://github.com/vais-ral/CCPi-Regularisation-Toolkit/blob/use_ctypes/demos/demo_gpu_regularisers3D.py
- [ ] opened https://github.com/vais-ral/CCPi-Regularisation-Toolkit/issues/186 to be fixed later
- [ ] `pyproject.toml` ~~generated by CMake with appropriate version string.~~ postpone as depends on https://github.com/vais-ral/CCPi-Regularisation-Toolkit/issues/185

### New functionality
In python it is possible now to pass to each function an `out` parameter for the regulariser to store the result. This was earlier not possible and the output was always allocated in the Cython layer. The output is also always returned, remaining backward compatible.

#### `infovector`
The `infovector`, a vector of 2 floats with information on whether the algorithm stopped having reached a certain tolerance, is never returned. To access the `infovector` the user needs to pass the function a preallocated vector with the key-worded parameter `infovector`.

#### `device`
Most of the functions have a OpenMP and CUDA version (CPU and GPU). The user could earlier specify which one to use with the positional parameter `device`. To remove lots of boilerplate code, `device` is now a key-worded argument because the C functions are wrapped in Python by the function below and, the positional parameters of each function are different, so I am never sure where `device` really is.

At any rate I believe this change adds clarity and the default is `device="cpu"`.

https://github.com/vais-ral/CCPi-Regularisation-Toolkit/blob/739e52af5542b2debb94ddfab0bce0901e7f2aae/src/Python/ccpi/filters/regularisers.py#L19-L32

#### PD_TV positional parameter order

I changed the parameter order of the `PD_TV` to reflect the underlying C/CUDA implementation.

### CODE REVIEW

**Comment by dkazanc** [comment]
thanks @paskino for looking into this. I think dropping Cython in favour of Ctypes is probably for the best here. Happy to contribute at some point, I was thinking of integrating more rigorous unit testing as well. 

We are planning to use some CuPy wrapped functions in production from the toolkit, so it is worth looking into how both cupy and cuda parts can be added in one conda build. I can do that later when the build is actually works. cheers

File: src/Python/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# pyproject.toml
+[build-system]
+requires = ["setuptools"]
+build-backend = "setuptools.build_meta"
+
+[project]
+version = "24.0.0"
+name='ccpi',
+name ='CCPi Core Imaging Library - Image regularisers',
```
Comment (casperdcl) [inline src/Python/pyproject.toml:9]:
```suggestion
name='ccpi'
```

File: test/test_run_test.py [hunk 2]
```diff
@@ -1,901 +1,858 @@
...
+
+        print ("#############PD TV CPU####################")
+        start_time = timeit.default_timer()
+        pd_cpu = PD_TV(pars['input'],
+                      pars['regularisation_parameter'],
+                      pars['number_of_iterations'],
+                      pars['tolerance_constant'],
+                      pars['methodTV'],
+                      pars['nonneg'],
+                      pars['lipschitz_const'], device='cpu')
```
Comment (paskino) [inline test/test_run_test.py:206]:
```suggestion
                      pars['lipschitz_const'], 
                      pars['methodTV'],
                      pars['nonneg'],
                      device='cpu')
```

File: test/test_run_test.py [hunk 3]
```diff
@@ -1,901 +1,858 @@
...
+
+        print ("##############PD TV GPU##################")
+        start_time = timeit.default_timer()
+        pd_gpu = PD_TV(pars['input'],
+              pars['regularisation_parameter'],
+              pars['number_of_iterations'],
+              pars['tolerance_constant'],
+              pars['methodTV'],
+              pars['nonneg'],
+              pars['lipschitz_const'], device='gpu')
```
Comment (paskino) [inline test/test_run_test.py:223]:
```suggestion
              pars['lipschitz_const'], 
              pars['methodTV'],
              pars['nonneg'],
              device='gpu')
```

**Review by casperdcl — APPROVED** [review]

**Pull request merged**